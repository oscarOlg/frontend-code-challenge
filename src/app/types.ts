export interface UserListComponentProps {
    usersData?: UserData;
    hasError: boolean;
    isDataLoading: boolean;
    resetData: () => void;
    handleOnDeleteClick: (id: string) => void;
    handleSearchByCountry: (country: string) => void;
}

export interface FilterBarComponentProps {
    onChangeColor: () => void;
    onToggleCountrySort: () => void;
    resetData: () => void;
    handleSearchByCountry: (country: string) => void;
}

export interface UserData {
    results: User[];
}

export interface User {
    name: Name;
    location: Location; 
    picture: Picture;
    login: Login;
}

export interface Name {
    title: string;
    first: string;
    last: string;
}

export interface Location {
    country: string;
}

export interface Picture {
    large: string;
    medium: string;
    thumbnail: string;
}

export interface Login {
    uuid: string;
}