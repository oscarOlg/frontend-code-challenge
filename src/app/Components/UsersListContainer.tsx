import React, { useState } from 'react';
import { User, UserListComponentProps, FilterBarComponentProps } from '../types';
import { FilterBarComponent } from './FilterBarComponent';
import Image from 'next/image';

import './styles/UserListContainer.css';

export const UsersListContainer = (props: UserListComponentProps) => {
  const { usersData, isDataLoading } = props;
  const [toggleColor, setToggleColor] = useState(false);
  const [toggleCountrySort, setCountryToggleSort] = useState(false);
  
  const defaultColor = { 
    color: 'black',
    borderColor: 'black',
    backgroundColor: 'white'
  };

  const customColorColor = { 
    color: 'white',
    borderColor: 'white',
    backgroundColor: 'rgb(78, 78, 78)'
  };

  const onChangeTableColor = () => {
    setToggleColor(prevValue => !prevValue);
  }

  const onToggleCountrySort = () => {
    setCountryToggleSort(prevValue => !prevValue);
  } 

  const sortByString = (a: string, b: string) => {
    let fa = a.toLowerCase(),
        fb = b.toLowerCase();

    if (fa < fb) {
        return -1;
    }
    if (fa > fb) {
        return 1;
    }
    return 0;
  }

  const sortDataByCountry = () => {
    return (
      [...usersData!.results].sort((userA, userB)=> sortByString(userA.location.country, userB.location.country))
    );
  }


  const FilterBarProps: FilterBarComponentProps = {
    onChangeColor: onChangeTableColor,
    onToggleCountrySort,
    handleSearchByCountry: props.handleSearchByCountry,
    resetData: props.resetData,
  }

  return (
    <div className='users-page-container'>
      <FilterBarComponent {...FilterBarProps} />
      { isDataLoading ? (
          <div data-testid='loading-title' className='loading-title'>Loading...</div> 
        ) : (
          !props.hasError ?
          <div data-testid='users-data-table' className='users-table'>
            <div className='users-table-headers'>
              <div className='user-table-cell'>Avatar</div>
              <div className='user-table-cell'>Name</div>
              <div className='user-table-cell'>Last Name</div>
              <div className='user-table-cell'>Country</div>
              <div className='user-table-cell'>Actions</div>
            </div>
            <div className='user-table-rows' style={!toggleColor ? defaultColor : customColorColor}>
              { usersData && 
                (toggleCountrySort ? sortDataByCountry() : usersData.results).map((user: User) => {
                  return (
                    <div key={user.login.uuid} className='user-table-row'>
                      <div className='user-table-cell'>
                        <Image 
                          src={`${user.picture.thumbnail}`} 
                          alt={`${user.name.first} ${user.name.last} avatar`} 
                          width={25}
                          height={25}
                        />
                      </div>
                      <div className='user-table-cell'>{user.name.first}</div>
                      <div className='user-table-cell'>{user.name.last}</div>
                      <div className='user-table-cell'>{user.location.country}</div>
                      <div className='user-table-cell'>
                        <button data-testid={`delete-button-${user.login.uuid}`} onClick={() => props.handleOnDeleteClick(user.login.uuid)}>Delete</button>
                      </div>
                    </div>
                  );
                })
              }  
            </div>
          </div>
          : (
            <div data-testid='error-title' className='error-title'>
              <h1>Error retriving users data</h1>
            </div>
          )
        )        
      }

    </div>
  )
}
