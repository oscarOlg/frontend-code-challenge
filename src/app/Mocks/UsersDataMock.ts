import { UserData } from "../types";

export const userDataMock: UserData = {
  results: [
    {
      name: {
          title: "Mrs",
          first: "Roseni",
          last: "de Souza"
      },
      location: {
          country: "Brazil",

      },
      login: {
          uuid: "abc123",
      },
      picture: {
        large: "https://randomuser.me/api/portraits/women/33.jpg",
        medium: "https://randomuser.me/api/portraits/med/women/33.jpg",
        thumbnail: "https://randomuser.me/api/portraits/thumb/women/33.jpg"
      },
    }
  ]
}