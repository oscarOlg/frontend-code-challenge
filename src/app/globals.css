
* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
  font-family: Verdana, Geneva, Tahoma, sans-serif;
}

html,
body {
  max-width: 100vw;
  overflow-x: hidden;
}

body {
  background: white;
}


button {
  padding: 5px 10px;
  border-radius: 5px;
  color: rgb(255, 255, 255);
  background-color: rgb(0, 101, 148);
}

input {
  border-radius: 5px;
  border-color: rgb(0, 101, 148);
}

input::placeholder {
  font-weight: bold;
  padding: 5px;
  opacity: 0.65;
  color: rgb(0, 101, 148);
}