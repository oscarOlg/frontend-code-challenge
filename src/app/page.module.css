.main {
  display: flex;
  flex-direction: column;
  gap: 30px;
  align-items: center;
  padding: 50px 0;
  min-height: 100vh;
}
