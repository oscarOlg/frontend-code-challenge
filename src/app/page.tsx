"use client";
import styles from './page.module.css'
import { UsersListContainer } from './Components/UsersListContainer';
import { useState, useEffect } from 'react';
import { UserData, UserListComponentProps } from './types';

export const Home = () => {
  const [usersData, setUsersData] = useState<UserData>({results: []});
  const [defaultData, setDefaultData] = useState<UserData>({results: []});
  const [filteredData, setFilteredData] = useState<UserData>({results: []});
  const [isDataLoading, setIsDataLoading] = useState(false); 
  const [error, setError] = useState(false);

  useEffect(() => {
    setFilteredData(usersData)
  }, [usersData])
  

  const resetData = () => {
    setUsersData(defaultData);
    setFilteredData(defaultData);
  }

  const handleOnDeleteClick = (id: string) => {
    const newUsers = filteredData.results.filter(user => user.login.uuid !== id);
    setFilteredData({results: newUsers});
  }

  const handleSearchByCountry = (country: string) => {
    if(country.length) { 
      const normCountry = country.toLocaleLowerCase();
      const filteredData = [...usersData.results].filter(user => user.location.country.toLowerCase().match(normCountry));
      setFilteredData({results: filteredData});
    } else {
      setFilteredData({results: []});
    }
  }

  const UserListProps: UserListComponentProps = {
    usersData: filteredData.results.length ? filteredData : usersData,
    isDataLoading, 
    resetData,
    handleOnDeleteClick,
    handleSearchByCountry,
    hasError: error,
  }

  const fetchUsers = (numOfUsers: number) => {
    fetch(`https://randomuser.me/api/?results=${numOfUsers}`)
      .then(response =>{
        if(response.ok) {
          return response.json();
        }
        throw response;
      })
      .then(data => {
        setUsersData(data);
        setDefaultData(data);
      })
      .catch(error => {
        console.error(error);
        setError(true);
      })
      .finally(() => {
        setIsDataLoading(false);
      })
  }

  useEffect(() => {
    setIsDataLoading(true);
    fetchUsers(100);
  }, []);

  return (
    <main className={styles.main}>
      <h1>Users list</h1>
      {<UsersListContainer {...UserListProps} />} 
    </main>
  )
}
