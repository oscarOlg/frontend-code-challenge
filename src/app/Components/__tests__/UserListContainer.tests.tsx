import '@testing-library/jest-dom';
import {render, fireEvent} from '@testing-library/react'
import { UsersListContainer } from '../UsersListContainer';
import { UserListComponentProps } from '@/app/types';
import { userDataMock } from '../../Mocks/UsersDataMock';


describe('UserListContainer tests', () => { 
    let props: UserListComponentProps;

    const renderComponent  = (props: UserListComponentProps)  => {
        return render(<UsersListContainer {...props} />)
    };

    let component: ReturnType<typeof renderComponent>;


    beforeEach(() => {
        props = {
            usersData: userDataMock,
            hasError: false,
            isDataLoading: false,
            resetData: () => {},
            handleOnDeleteClick: (id: string) => {},
            handleSearchByCountry: (country: string) => {},
        };
    });
      
    afterEach(() => {
        if (component) {
          component.unmount();
        }
    });
    
    
    it('Should render UserListContainer correctly when there is data', () => {
        component = renderComponent(props);
        const usersTable = component.queryByTestId('users-data-table');
        expect(usersTable).toBeVisible();
        expect(component.baseElement).toMatchSnapshot();
    });

    it('Should render Loading title correctly when data is loading', () => {
        props.isDataLoading = true;
        component = renderComponent(props);
        const loadingTitle = component.queryByTestId('loading-title');
        expect(loadingTitle).toBeVisible();
    });

    it('Should render error title correctly when there is an error when fetching data', () => {
        props.hasError = true;
        component = renderComponent(props);
        const errorTitle = component.queryByTestId('error-title');
        expect(errorTitle).toBeVisible();
    });

    it('Should handle delete correctly when deleting an User', () => {
        props.handleOnDeleteClick = jest.fn()
        component = renderComponent(props);
        const deleteButton = component.queryByTestId(`delete-button-${userDataMock.results[0].login.uuid}`) as HTMLButtonElement;
        expect(deleteButton).toBeVisible();
        fireEvent.click(deleteButton);
        expect(props.handleOnDeleteClick).toHaveBeenCalled();
    });

});