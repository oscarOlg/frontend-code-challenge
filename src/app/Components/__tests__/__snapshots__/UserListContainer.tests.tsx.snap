// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListContainer tests Should render UserListContainer correctly when there is data 1`] = `
<body>
  <div>
    <div
      class="users-page-container"
    >
      <div
        class="filterbar-container"
        data-testid="filter-bar"
      >
        <button
          data-testid="change-color-button"
        >
          Color table
        </button>
        <button
          data-testid="sort-button"
        >
          Sort by country
        </button>
        <button
          data-testid="reset-button"
        >
          reset
        </button>
        <input
          data-testid="search-input"
          placeholder="Search by Country"
        />
      </div>
      <div
        class="users-table"
        data-testid="users-data-table"
      >
        <div
          class="users-table-headers"
        >
          <div
            class="user-table-cell"
          >
            Avatar
          </div>
          <div
            class="user-table-cell"
          >
            Name
          </div>
          <div
            class="user-table-cell"
          >
            Last Name
          </div>
          <div
            class="user-table-cell"
          >
            Country
          </div>
          <div
            class="user-table-cell"
          >
            Actions
          </div>
        </div>
        <div
          class="user-table-rows"
          style="color: black; border-color: black; background-color: white;"
        >
          <div
            class="user-table-row"
          >
            <div
              class="user-table-cell"
            >
              <img
                alt="Roseni de Souza avatar"
                data-nimg="1"
                decoding="async"
                height="25"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Frandomuser.me%2Fapi%2Fportraits%2Fthumb%2Fwomen%2F33.jpg&w=64&q=75"
                srcset="/_next/image?url=https%3A%2F%2Frandomuser.me%2Fapi%2Fportraits%2Fthumb%2Fwomen%2F33.jpg&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Frandomuser.me%2Fapi%2Fportraits%2Fthumb%2Fwomen%2F33.jpg&w=64&q=75 2x"
                style="color: transparent;"
                width="25"
              />
            </div>
            <div
              class="user-table-cell"
            >
              Roseni
            </div>
            <div
              class="user-table-cell"
            >
              de Souza
            </div>
            <div
              class="user-table-cell"
            >
              Brazil
            </div>
            <div
              class="user-table-cell"
            >
              <button
                data-testid="delete-button-abc123"
              >
                Delete
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
